ch5/ex-5-17-qsort-fields.c

```diff
@@ -0,0 +1,228 @@
+ *
+ * Add a field-handling capability, so sorting may be done on fields within
+ * lines, each field sorted according to an independent set of options.  (The
+ * index for this book was sorted with -df for the index category and -n for the
+ * page numbers.)
+ */
+
+#include <ctype.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include "lines.h"
+#include "quicksort.h"
+
+#define CHARCOMP int (*)(char *, char *)
+#define VOIDCOMP int (*)(void *, void *)
+
+void parseargs(int argc, char *argv[]);
+char **getsortdata(char *line);
+int compare(char **s1, char **s2);
+
+char *line(char *sortdata[]);
+char *fieldspace(char *sortdata[]);
+void fold(char *value);
+void dirsort(char *value);
+int strcmpr(char *, char *);
+int numcmp(char *, char *);
+int numcmpr(char *, char *);
+void freestuff(char ***linedata, int nlines);
+
+int *sortidxs; /* indexes of fields to sort by, ordered by priority */
+int nsortflds; /* number of fields we're sorting by */
+int maxsortidx; /* highest index of the fields we're sorting by */
+int (**compares)(char *, char *); /* comparison function for each sort field */
+int *folds; /* whether to fold each sort field */
+int *dirsorts; /* whether to use dir sort for each sort field */
+
+int main(int argc, char *argv[])
+{
+	int i, nlines;
+	char *buff, **lines, ***linedata;
+
+	parseargs(argc, argv);
+
+	if ((nlines = readlines(&buff, &lines)) == LNS_ERROR) {
+		printf("input too big to sort\n");
+		return 0;
+	}
+
+	linedata = malloc(nlines * sizeof(char **));
+	for (i = 0; i < nlines; i++)
+		linedata[i] = getsortdata(lines[i]);
+
+	quicksort((void **)linedata, 0, nlines - 1, (VOIDCOMP)compare);
+
+	for (i = 0; i < nlines; i++)
+		lines[i] = line(linedata[i]);
+
+	writelines(lines, nlines);
+
+	freestuff(linedata, nlines);
+	freelines(buff, lines);
+	return 0;
+}
+
+int compare(char **sortdata1, char **sortdata2)
+{
+	int i, rslt = 0;
+
+	for (i = 0; i < nsortflds && rslt == 0; i++)
+		rslt = (*compares[i])(sortdata1[i], sortdata2[i]);
+	return rslt;
+}
+
+void parseargs(int argc, char *argv[])
+{
+	int i;
+	int maxsortcount = argc - 1;
+
+	nsortflds = 0;
+	sortidxs = malloc(maxsortcount * sizeof(int));
+	compares = malloc(maxsortcount * sizeof(CHARCOMP));
+	folds = malloc(maxsortcount * sizeof(int));
+	dirsorts = malloc(maxsortcount * sizeof(int));
+
+	argv++;
+	for (i = 0; i < argc - 1; i++) {
+		char *arg = argv[i];
+
+		int fieldidx;
+		int numeric = 0;
+		int reverse = 0;
+		int (*compare)(char *, char *);
+		int fold = 0;
+		int dirsort = 0;
+
+		if (*arg == '-' && isdigit(*++arg)) {
+			fieldidx = atoi(arg) - 1;
+			while (*++arg) {
+				if (*arg == 'n')
+					numeric = 1;
+				else if (*arg == 'r')
+					reverse = 1;
+				else if (*arg == 'f')
+					fold = 1;
+				else if (*arg == 'd')
+					dirsort = 1;
+			}
+			if (numeric && reverse)
+				compare = numcmpr;
+			else if (numeric)
+				compare = numcmp;
+			else if (reverse)
+				compare = strcmpr;
+			else
+				compare = (CHARCOMP)strcmp;
```

Comment:
I don't think you need this cast? Does removing it cause an error?

Reply:
Removing it causes a compiler warning.  (char *, char *) vs (const char *, const char *).  The natural solution is probably to make all the compare signatures (const char *, ...), but I don't think that int the knr toolbox.